### PR TITLE
feat: MN3 and MilkyWayPotential2022

### PR DESCRIPTION
**Pull request by adrn**
This implements the MN3 and MilkyWayPotential2022 potential models.

In writing the implementations, I have a few questions:
1. Subclassing `AbstractPotential` seems to require specifying several boilerplate class attributes:
  ```python
  _: KW_ONLY
  units: AbstractUnitSystem = eqx.field(converter=unitsystem, static=True)
  constants: ImmutableMap[str, Quantity] = eqx.field(
      default=default_constants, converter=ImmutableMap
  )
  ```
  Can you say more about these? Is there no way we can move default behavior to the `__check_init__` of the base class?

2. Why do the potential methods all have to be decorated with `@partial(jax.jit)`? Isn't that handled by equinox?
3. The tests are currently failing because I need to update the expected values for the two potentials I implemented. How did you compute these and how do you know they are "correct" or actually expected? In Gala, I did some tests against sympy -- do we want something like that in galax?

### CODE REVIEW

File: src/galax/potential/_potential/builtin/builtin.py [hunk 1]
```diff
@@ -394,6 +395,140 @@ def _potential(
...
+    [
+        [0.0036, -0.0330, 0.1117, -0.1335, 0.1749],
+        [-0.0131, 0.1090, -0.3035, 0.2921, -5.7976],
+        [-0.0048, 0.0454, -0.1425, 0.1012, 6.7120],
+        [-0.0158, 0.0993, -0.2070, -0.7089, 0.6445],
+        [-0.0319, 0.1514, -0.1279, -0.9325, 2.6836],
+        [-0.0326, 0.1816, -0.2943, -0.6329, 2.3193],
+    ]
+)
+_mn3_K_neg_dens = xp.asarray(
```
[the remark this answers is not in the document]
Comment (nstarman) [reply src/galax/potential/_potential/builtin/builtin.py:408]: Just remembered this was a thing.
🧐 we should be Finalizing a lot more constants. In principle it makes things faster because Python knows it's a constant. Maybe one day JAX will use python type annotations when jitting / JAX will play well with [mypyc](https://mypyc.readthedocs.io/en/latest/)

File: src/galax/potential/_potential/builtin/builtin.py [hunk 2]
```diff
@@ -394,6 +395,140 @@ def _potential(
...
+        ]
+
+    @partial(jax.jit)
+    def _potential(
+        self, q: gt.BatchQVec3, t: gt.BatchableRealQScalar, /
+    ) -> gt.SpecificEnergyBatchScalar:
+        return Quantity(
+            xp.sum(
+                xp.asarray(
+                    [mn.potential(q, t).value for mn in self._get_mn_components(t)]
```
Comment (nstarman) [inline src/galax/potential/_potential/builtin/builtin.py:507]: I think this might need to be 
```suggestion
                    [mn.potential(q, t).to_value(self.units["specific energy"]) for mn in self._get_mn_components(t)]
```
In which case store the out as an intermediate variable?

File: src/galax/potential/_potential/builtin/special.py [hunk 3]
```diff
@@ -38,7 +44,7 @@ def _parse_input_comp(
         return instance
 
     if units == dimensionless:
-        default = {k: v.value for k, v in default.items()}
+        default = {k: v.value if hasattr(v, "value") else v for k, v in default.items()}
```
Comment (nstarman) [inline src/galax/potential/_potential/builtin/special.py:47]: Should it be a `AbstractQuantity` check, or is an input Astropy Quantity a concern?

File: src/galax/potential/_potential/builtin/special.py [hunk 4]
```diff
@@ -58,7 +64,7 @@ class BovyMWPotential2014(AbstractCompositePotential):
     Parameters
     ----------
     units : `~gala.units.UnitSystem` (optional)
-        Set of non-reducable units that specify (at minimum) the
+        Set of non-reducible units that specify (at minimum) the
```
Comment (nstarman) [inline src/galax/potential/_potential/builtin/special.py:67]: :shipit:

File: src/galax/potential/_potential/builtin/special.py [hunk 5]
```diff
@@ -306,3 +309,93 @@ def __init__(
...
+    """
+
+    _data: dict[str, AbstractPotentialBase] = eqx.field(init=False)
+    _: KW_ONLY
+    units: AbstractUnitSystem = eqx.field(init=True, static=True, converter=unitsystem)
+    constants: ImmutableMap[str, Quantity] = eqx.field(
+        default=default_constants, converter=ImmutableMap
+    )
+
+    # TODO: as an actual potential instance, then use `replace`?
```
Comment (nstarman) [inline src/galax/potential/_potential/builtin/special.py:353]: Yeah, I'm leaning that way as well. Have to do testing with JIT / autodiff, though.

Comment (nstarman) [reply src/galax/potential/_potential/builtin/special.py:353]: If it is full instances, then `replace` might allow us to get rid of the annoying helper function `_parse_input_comp`

**Review by nstarman — COMMENTED** [review]
😎

File: src/galax/potential/_potential/builtin/builtin.py [hunk 6]
```diff
@@ -15,6 +15,7 @@
     "SatohPotential",
     "StoneOstriker15Potential",
     "TriaxialHernquistPotential",
+    "MN3ExponentialPotential",
 ]
 
 from dataclasses import KW_ONLY
```
[the remark this answers is not in the document]
Comment (adrn) [reply src/galax/potential/_potential/builtin/builtin.py:21]: Ah, already imported as `final`, so I'll implement locally and not take your suggestions below.

File: src/galax/potential/_potential/builtin/builtin.py [hunk 7]
```diff
@@ -394,6 +395,140 @@ def _potential(
...
+        [-0.0051, 0.0287, -0.0361, -0.0544, 0.2242],
+        [-0.0358, 0.2610, -0.6987, -0.1193, 2.0074],
+        [-0.0830, 0.4992, -0.7967, -1.2966, 4.4441],
+        [-0.0247, 0.1718, -0.4124, -0.5944, 0.7333],
+    ]
+)
+
+
+@final
+class MN3ExponentialPotential(AbstractPotential):
```
[the remark this answers is not in the document]
Comment (adrn) [reply src/galax/potential/_potential/builtin/builtin.py:421]: Well, they are both mixtures of 3 MN disks -- I don't have a strong opinion!

File: src/galax/potential/_potential/builtin/special.py (hunk 3 above)
Comment (adrn) [reply src/galax/potential/_potential/builtin/special.py:47]: Wouldn't an Astropy quantity get converted in the `cls()` call below?

File: src/galax/potential/_potential/builtin/special.py (hunk 5 above)
Comment (adrn) [reply src/galax/potential/_potential/builtin/special.py:353]: For future!

**Comment by adrn** [comment]
I sloppily let a merge commit sneak in but maybe that is ok if you squash and merge?

File: src/galax/potential/_potential/builtin/special.py (hunk 3 above)
Comment (nstarman) [reply src/galax/potential/_potential/builtin/special.py:47]: I believe so.
This was more about whether the explicit ``{k: v.value if isinstance(v, AbstractQuantity)...}`` would work here.

File: src/galax/potential/_potential/builtin/builtin.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (adrn) [reply src/galax/potential/_potential/builtin/builtin.py:421]: > A goal is to avoid multi-modality
Haha tell that to `SkyCoord` 😂 -- but ok, sounds good.

File: src/galax/potential/_potential/builtin/special.py (hunk 3 above)
Comment (adrn) [reply src/galax/potential/_potential/builtin/special.py:47]: Actually, thinking about this function itself: does it make sense? Should we ever let someone pass in a dimensionless unit system to these classes? And if so, do we want to just strip the units from the standard default values? That seems strange to me...

File: src/galax/potential/_potential/builtin/builtin.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nstarman) [reply src/galax/potential/_potential/builtin/builtin.py:421]: > Haha tell that to SkyCoord

Totally. Agree.

[the remark this answers is not in the document]
Comment (adrn) [reply src/galax/potential/_potential/builtin/builtin.py]: OK split!

File: src/galax/potential/_potential/builtin/builtin.py [hunk 8]
```diff
@@ -9,18 +9,19 @@
...
     "SatohPotential",
     "StoneOstriker15Potential",
     "TriaxialHernquistPotential",
-    "MN3ExponentialPotential",
 ]
 
 from dataclasses import KW_ONLY
 from functools import partial
-from typing import Any, final
+from typing import Any, Final, final
```
Comment (nstarman) [inline src/galax/potential/_potential/builtin/builtin.py:24]:
```suggestion
from typing import Any, Final, final
from jaxtyping import Array
```

File: src/galax/potential/_potential/builtin/builtin.py [hunk 9]
```diff
@@ -534,6 +511,55 @@ def _density(
...
+    As described in the above reference, this approximation has two options: (1)
+    with the ``positive_density=True`` argument set, this density will be
+    positive everywhere, but is only a good approximation of the exponential
+    density within about 5 disk scale lengths, and (2) with
+    ``positive_density=False``, this density will be negative in some regions,
+    but is a better approximation out to about 7 or 8 disk scale lengths.
+    """
+
+    @property
+    def _b_coeffs(self) -> Any:
```
Comment (nstarman) [inline src/galax/potential/_potential/builtin/builtin.py:559]:
```suggestion
    def _b_coeffs(self) -> Array:
```

File: src/galax/potential/_potential/builtin/builtin.py [hunk 10]
```diff
@@ -534,6 +511,55 @@ def _density(
...
```
Comment (nstarman) [inline src/galax/potential/_potential/builtin/builtin.py:535]:
```suggestion
    def _b_coeffs(self) -> Array:
```

File: src/galax/potential/_potential/builtin/nfw.py
Comment (nstarman) [inline src/galax/potential/_potential/builtin/nfw.py:1]: 👍. Great to fill out the `from_X`.

File: src/galax/potential/_potential/builtin/builtin.py [hunk 11]
```diff
@@ -394,6 +396,173 @@ def _potential(
...
+    constants: ImmutableMap[str, Quantity] = eqx.field(
+        default=default_constants, converter=ImmutableMap
+    )
+
+    def _get_mn_components(
+        self, t: gt.BatchableRealQScalar, /
+    ) -> list[MiyamotoNagaiPotential]:
+        hR = self.h_R(t)
+        hzR = (self.h_z(t) / hR).decompose(dimensionless).value
+        K = _mn3_K_pos_dens if self.positive_density else _mn3_K_neg_dens
```
Comment (nstarman) [inline src/galax/potential/_potential/builtin/builtin.py:457]: jax.lax.switch?
There are so many places across the code base (my fault!) where we should be AB testing canonical Python vs JAX syntax

**Review by nstarman — COMMENTED** [review]
Nice! Seeing the separated classes convinced me that this is the right approach.

File: src/galax/potential/_potential/builtin/special.py (hunk 3 above)
Comment (adrn) [reply src/galax/potential/_potential/builtin/special.py:47]: I was more lenient in Gala about letting potentials take a dimensionless unit system for users that didn't want to use units, but it did lead to some weird hidden behavior (like any mass parameter was actually GM because I set G=1 for dimensionless cases, and that wasn't documented well). I guess I'm ok with not allowing this anymore as long as we make some shorthands for easily defining simulation unit systems in unxt. A common one is to have a velocity scale, a length scale, and G=1, for example.

Either way, I don't think any potentials with default parameters (like the special's) should be allowed to be dimensionless.

File: src/galax/potential/_potential/builtin/builtin.py [hunk 12]
```diff
@@ -394,6 +395,140 @@ def _potential(
...
+        if self.sech2_z:
+            b_hR = -0.033 * hzR**3 + 0.262 * hzR**2 + 0.659 * hzR
+        else:
+            b_hR = -0.269 * hzR**3 + 1.08 * hzR**2 + 1.092 * hzR
+
+        x = jnp.vander(b_hR[None], N=5)[0]
+        param_vec = K @ x
+
+        # use fitting function to get the Miyamoto-Nagai component parameters
+        # NOTE(adrn): I had to add .value here otherwise when @jit'ing, this function
```
[the remark this answers is not in the document]
Comment (adrn) [reply src/galax/potential/_potential/builtin/builtin.py:466]: I'm stumped on this one!

File: src/galax/potential/_potential/builtin/builtin.py (hunk 11 above)
Comment (adrn) [reply src/galax/potential/_potential/builtin/builtin.py:458]: Shouldn't this be ok as long as `positive_density` is static? No gradients should be taken w.r.t. any of these variables.

File: src/galax/potential/_potential/builtin/special.py (hunk 3 above)
Comment (nstarman) [reply src/galax/potential/_potential/builtin/special.py:47]: SGTM.

File: src/galax/potential/_potential/builtin/builtin.py (hunk 11 above)
Comment (nstarman) [reply src/galax/potential/_potential/builtin/builtin.py:458]: I think that's right.

File: src/galax/potential/_potential/builtin/builtin.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (nstarman) [reply src/galax/potential/_potential/builtin/builtin.py:467]: Is it the `zip`? 
if you tried 3 MNs explicitly, maybe it wouldn't stall out?

File: src/galax/potential/_potential/builtin/special.py (hunk 3 above)
Comment (adrn) [reply src/galax/potential/_potential/builtin/special.py:47]: I'll do this in a separate PR - see https://github.com/GalacticDynamics/unxt/issues/184 and #438.

File: src/galax/potential/_potential/builtin/builtin.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (adrn) [reply src/galax/potential/_potential/builtin/builtin.py]: Yea that seems to have fixed it! Also probably better form anyways, since we know there have to be 3 components in this class.

**Comment by adrn** [comment]
I think the only remaining issue here is to address the "low" test coverage of the added potentials.

**Comment by adrn** [comment]
@nstarman I think this is ready for another look!

**Review by nstarman — APPROVED** [review]
🎉
